nemo_curator/filters/synthetic.py

```diff
@@ -0,0 +1,219 @@
+        self.nim_model = nim_model
+        self.percentile = percentile
+        if truncate:
+            self.truncate = truncate
+        try:
+            self.client = OpenAI(base_url=self.base_url, api_key=self.api_key)
+        except Exception as e:
+            print(f"Error accessing NIM model: {e}")
+        self.batch_size = batch_size
+        self.text_fields = text_fields
+
+    @batched
+    def score_document(self, df: pd.DataFrame):
+
+        document_score = self._calc_similarity_nim(
+            df[self.text_fields[0]].to_list(), df[self.text_fields[1]].to_list()
+        )
+        return pd.Series(document_score, index=df.index)
+
+    @batched
+    def keep_document(self, scores: pd.Series):
+        filter_threshold = np.percentile(scores, self.percentile)
+        return scores <= filter_threshold
+
+    def _get_nim_embedding(self, text, input_type):
+        # Obtain embeddings from nim model
+        if isinstance(text, list):
+            input_ = text
+        elif isinstance(text, str):
+            input_ = [text]
+
+        try:
+            response = self.client.embeddings.create(
+                input=input_,
+                model=self.nim_model,
+                encoding_format="float",
+                extra_body={"input_type": input_type, "truncate": self.truncate},
+            )
+        except Exception as e:
+            print(f"Error: {e}")
+            response = None
+
+        if response:
+            if isinstance(text, list):
+                embeddings = [r.embedding for r in response.data]
+            elif isinstance(text, str):
+                embeddings = response.data[0].embedding
+            return embeddings
+        else:
+            return []
+
+    def _calc_similarity_nim(self, context, question):
+        # cosine similarity
+        doc_embed = self._get_nim_embedding(text=context, input_type="passage")
+        q_embed = self._get_nim_embedding(text=question, input_type="query")
+        if isinstance(context, list) and isinstance(question, list):
+            if doc_embed and q_embed:
+                sim = np.diag(np.dot(np.array(doc_embed), np.array(q_embed).T))
+            else:
+                sim = np.zeros(len(context))
+        else:
+            if doc_embed and q_embed:
+                sim = np.dot(doc_embed, q_embed)
+            else:
+                sim = 0.0
+
+        return sim
+
+    def __dask_tokenize__(self):
+        return normalize_token(EasinessFilter)
+
+
+# ----------------------------------------------------------------------------80
+# ----------------------- Answerability Filter ---------------------------------
+# ----------------------------------------------------------------------------80
+
+
+class AnswerabilityFilter(DocumentFilter):
+    """
+    Discards questions that are not answerable by content present in the
+    context document
+    """
+
+    def __init__(
+        self,
+        base_url: str = None,
```

Comment:
same comment about defaulting when we shouldn't

Reply:
fixed